### PR TITLE
Prealloc tests

### LINKED ISSUE #46: Writing unit tests for Preallocator
I'm not entirely sure about the Preallocator implementation so I wasn't able to do it, but we definitely need this.

I already created a boilerplate at `tests/testPreallocator.py`

### PR DESCRIPTION
**Pull request by akhilnadigatla**
Fixes #46.

Changes proposed in this PR:
- Added tests for the following methods in the Preallocator object:
    1. Obtaining the size of machine pool associated with a VM name
    2. Updating the number of machines in the pool for a VM name
    3. (Pre)Allocating a VM to run a job
    4. Freeing a VM to return to a pool
    5. Adding a VM
    6. Removing a VM
    7. Obtaining IDs for machines in the pools
    8. 'Creating' a VM for the preallocator to manage
    9. 'Removing' a VM from the preallocator
    10. Get a summary of the machines in a pool

### CODE REVIEW

**Comment by fanpu** [comment]
Hi @akhilnadigatla, thank you for your contribution! I am running the test with `python -m unittest tests/testPreallocator.py` (correct me if this was not intended). I'm obtaining the following results from the tests right now:

```
FEEEEFE.FFFE
======================================================================
ERROR: test_allocVM (tests.testPreallocator.TestPreallocator)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/fanpu/Tango/tests/testPreallocator.py", line 61, in test_allocVM
    self.preallocator.update(self.vm, 0)
  File "/home/fanpu/Tango/preallocator.py", line 67, in update
    threading.Thread(target=self.__destroy(vm)).start()
  File "/home/fanpu/Tango/preallocator.py", line 174, in __destroy
    dieVM = self.machines.get(vm.name)[1].get_nowait()
  File "/home/fanpu/Tango/tangoObjects.py", line 249, in get_nowait
    return self.get(False)
  File "/home/fanpu/Tango/tangoObjects.py", line 244, in get
    item = pickle.loads(item)
TypeError: a bytes-like object is required, not 'NoneType'

======================================================================
ERROR: test_create (tests.testPreallocator.TestPreallocator)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/fanpu/Tango/tests/testPreallocator.py", line 102, in test_create
    self.preallocator.__create(self.vm, 7)
AttributeError: 'Preallocator' object has no attribute '_TestPreallocator__create'

======================================================================
ERROR: test_createVM (tests.testPreallocator.TestPreallocator)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/fanpu/Tango/tests/testPreallocator.py", line 113, in test_createVM
    self.preallocator.createVM(self.vm)
  File "/home/fanpu/Tango/preallocator.py", line 187, in createVM
    vmms = self.vmms[vm.vmms]
KeyError: None

======================================================================
ERROR: test_destroy (tests.testPreallocator.TestPreallocator)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/fanpu/Tango/tests/testPreallocator.py", line 107, in test_destroy
    self.preallocator.__destroy(self.vm)
AttributeError: 'Preallocator' object has no attribute '_TestPreallocator__destroy'

======================================================================
ERROR: test_freeVM (tests.testPreallocator.TestPreallocator)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/fanpu/Tango/tests/testPreallocator.py", line 74, in test_freeVM
    self.preallocator.freeVM(vm)
  File "/home/fanpu/Tango/preallocator.py", line 106, in freeVM
    vmms = self.vmms[vm.vmms]
AttributeError: 'NoneType' object has no attribute 'vmms'

======================================================================
ERROR: test_update (tests.testPreallocator.TestPreallocator)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/fanpu/Tango/tests/testPreallocator.py", line 52, in test_update
    self.preallocator.update(self.vm, 10)
  File "/home/fanpu/Tango/preallocator.py", line 59, in update
    threading.Thread(target=self.__create(vm, delta)).start()
  File "/home/fanpu/Tango/preallocator.py", line 149, in __create
    vmms = self.vmms[vm.vmms]
KeyError: None

======================================================================
FAIL: test_addVM (tests.testPreallocator.TestPreallocator)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/fanpu/Tango/tests/testPreallocator.py", line 85, in test_addVM
    self.assertEqual(self.preallocator.poolSize(self.vm), prevSize + 1)
AssertionError: 0 != 1

======================================================================
FAIL: test_destroyVM (tests.testPreallocator.TestPreallocator)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/fanpu/Tango/tests/testPreallocator.py", line 126, in test_destroyVM
    self.assertEqual(res, 0)
AssertionError: -1 != 0

======================================================================
FAIL: test_getPool (tests.testPreallocator.TestPreallocator)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/fanpu/Tango/tests/testPreallocator.py", line 145, in test_getPool
    self.assertTrue(pool["total"] == [])
AssertionError: False is not true

======================================================================
FAIL: test_poolSize (tests.testPreallocator.TestPreallocator)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/fanpu/Tango/tests/testPreallocator.py", line 44, in test_poolSize
    self.assertEqual(self.preallocator.poolSize(self.vm.name), 0)
AssertionError: 1 != 0

======================================================================
FAIL: test_removeVM (tests.testPreallocator.TestPreallocator)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/fanpu/Tango/tests/testPreallocator.py", line 91, in test_removeVM
    self.assertEqual(self.preallocator.poolSize(self.vm), prevSize - 1)
AssertionError: 0 != -1

----------------------------------------------------------------------
Ran 12 tests in 0.006s

FAILED (failures=5, errors=6)
```

If this was meant to be run in another way, please let me know

**Comment by fanpu** [comment]
The VMMS config that it was being tested with is 'localDocker'

File: tests/testPreallocator.py [hunk 1]
```diff
@@ -1,51 +1,156 @@
...
+        elif Config.VMMS_NAME == "distDocker":
+            from vmms.distDocker import DistDocker
+            vmms = DistDocker()
+            self.preallocator = Preallocator({"distDocker": vmms})
+
+        else:
+            vmms = None
+            self.preallocator = Preallocator({"default": vmms})
+       
+        self.vm = TangoMachine()
```
Comment (fanpu) [inline tests/testPreallocator.py:36]: I think many of the errors are stemming from the fact that `vmms` was not passed into the TangoMachine here, but there are still failures after this change is made

**Review by fanpu — CHANGES_REQUESTED** [review]

File: tests/testPreallocator.py (hunk 1 above)
Comment (akhilnadigatla) [reply tests/testPreallocator.py:36]: Hi @fanpu. Thank you for this information! I interpreted something else as a successful run of these tests. I will look into the issues with this code.

Comment (fanpu) [reply tests/testPreallocator.py:36]: Please let me know if you run into any issues with the setup process for testing, https://docs.autolabproject.com/tango/ is a good reference

**Comment by mojojojo99** [comment]
Hey @fanpu I've fixed all the tests, now they should be working. The changes to the tests are as follows:

- I've set up the framework such that you can, in future, add to the list `self.test_machines` the names of the vmms to test. Currently, we will only test the most basic `localDocker`. There were previously issues in setting up the allocator and also properly creating a `TangoMachine` instance.
- I've also fixed and rewrote a couple tests: some of them were testing for private api which should be removed.

My suggestions:
- Ideally, we would be running the unittests with these mocks: I've created a sample of how we could potentially use them in #191, but I think these should suffice for now.

**Review by fanpu — APPROVED** [review]
Tests works and are very reasonable. Thanks for your contributions @akhilnadigatla  @mojojojo99 !

**Comment by fanpu** [comment]
Don't worry about testing the other vmms - they are really hard to stub/mock

**Comment by akhilnadigatla** [comment]
Awesome! Thanks, @mojojojo99 @fanpu!

**Pull request merged**